gensim/test/test_doc2vec.py

```diff
@@ -687,6 +687,22 @@ def testTrainWarning(self, l):
     def testLoadOnClassError(self):
         """Test if exception is raised when loading doc2vec model on instance"""
         self.assertRaises(AttributeError, load_on_instance)
+
+    def test_most_similiar_uses_newly_trained_vectors(self):
+        """Test if the doc2vec model deletes the normalized vectors that are used by most similar
+        when it is retrained"""
+        corpus = DocsLeeCorpus()
+        model = doc2vec.Doc2Vec(vector_size=10, min_count=1, workers=1)
+        model.build_vocab(corpus)
+
+        model.train(corpus, total_examples=model.corpus_count, epochs=1, start_alpha=0.1, end_alpha=0.1)
+        model.docvecs.most_similar(1)
```

Comment:
it's better to test in the same manner (i.e. using callbacks) as mentioned in https://github.com/RaRe-Technologies/gensim/issues/2260#issue-377788511 (i.e. call most_similar "during training", not after). Same things for Word2Vec and FastText

Reply:
I see your point. The problem is that above mentioned function `_clear_post_train`, which is responsible to reset the previously trained docvecs, is only called after the whole training is done. This means that calls of `most_similar` inside a callback, won't be able to access the newly trained vectors.
One solution would be to add a `replace=False` to `most_similiar` which is further passed down to `init_sims`. The user then has to change `replace` to `True` in order to be able to evaluate the newly calculated dimensions during training.
Further, this modification can be done without changing the current behavior.